packages/monitors/index.js

```diff
@@ -80,229 +88,280 @@ async function run({
+      const tokenBalanceClient = new TokenBalanceClient(
+        logger,
+        getAbi("ExpandedERC20"),
+        web3,
+        collateralTokenAddress,
+        syntheticTokenAddress
+      );
 
-    // 2. Balance monitor to inform if monitored addresses drop below critical thresholds.
-    const tokenBalanceClient = new TokenBalanceClient(
-      logger,
-      getAbi("ExpandedERC20"),
-      web3,
-      collateralTokenAddress,
-      syntheticTokenAddress
-    );
-
-    const balanceMonitor = new BalanceMonitor({
-      logger,
-      tokenBalanceClient,
-      monitorConfig,
-      financialContractProps
-    });
+      const balanceMonitor = new BalanceMonitor({
+        logger,
+        tokenBalanceClient,
+        monitorConfig,
+        financialContractProps
+      });
 
-    // 3. Collateralization Ratio monitor.
-    const financialContractClient = new FinancialContractClient(
-      logger,
-      getAbi(monitorConfig.contractType, monitorConfig.contractVersion),
-      web3,
-      financialContractAddress,
-      collateralDecimals,
-      syntheticDecimals,
-      medianizerPriceFeed.getPriceFeedDecimals(),
-      monitorConfig.contractType
-    );
-
-    const crMonitor = new CRMonitor({
-      logger,
-      financialContractClient,
-      priceFeed: medianizerPriceFeed,
-      monitorConfig,
-      financialContractProps
-    });
+      // 3. Collateralization Ratio monitor.
+      const financialContractClient = new FinancialContractClient(
+        logger,
+        getAbi(monitorConfig.contractType, monitorConfig.contractVersion),
+        web3,
+        financialContractAddress,
+        collateralDecimals,
+        syntheticDecimals,
+        medianizerPriceFeed.getPriceFeedDecimals(),
+        monitorConfig.contractType
+      );
 
-    // 4. Synthetic Peg Monitor.
-    const syntheticPegMonitor = new SyntheticPegMonitor({
-      logger,
-      web3,
-      uniswapPriceFeed: tokenPriceFeed,
-      medianizerPriceFeed,
-      denominatorPriceFeed,
-      monitorConfig,
-      financialContractProps
-    });
+      const crMonitor = new CRMonitor({
+        logger,
+        financialContractClient,
+        priceFeed: medianizerPriceFeed,
+        monitorConfig,
+        financialContractProps
+      });
+
+      // 4. Synthetic Peg Monitor.
+      const syntheticPegMonitor = new SyntheticPegMonitor({
+        logger,
+        web3,
+        uniswapPriceFeed: tokenPriceFeed,
+        medianizerPriceFeed,
+        denominatorPriceFeed,
+        monitorConfig,
+        financialContractProps
+      });
+
+      logger.debug({
+        at: "Monitor#index",
+        message: "Monitor initialized",
+        collateralDecimals: Number(collateralDecimals),
+        syntheticDecimals: Number(syntheticDecimals),
+        priceFeedDecimals: Number(medianizerPriceFeed.getPriceFeedDecimals()),
+        tokenPriceFeedConfig,
+        medianizerPriceFeedConfig,
+        monitorConfig
+      });
+
+      asyncLoopLogic = async () => {
```

Comment:
it kinda sucks that this needs to be repeated. would it not be posible to put the branching logic one "main" loop and then just choose which of the sets of promises to run?

Reply:
Yeah agreeed I actually tried this originally by having each branch push an array of promises. The problem is that some of the monitor functions area not asynchronous like the getEvent methods. Maybe i could turn those into promises by wrapping a 'return new Promise' around them

Reply:
Fixed, referenced in [this comment](https://github.com/UMAprotocol/protocol/pull/2597#discussion_r582891486)